Add unit test for the whole project

guilhermeferreira opened the pull request:
The branch improves the existing unit tests and adds news ones. The goal is to cover more than 90% of the library code. 

Coverage summary:
- 717 out of 756 lines covered (around 95% of all code lines).
- 142 out of 160 functions covered (around 90% functions/methods).

In order to pass all test cases, this branch requires the following fixes:
#20 
#24 
#41 
#44 
#42 
#48 
#45
#43 
#46 
#47 
#49 
#50

fpagliughi commented:
This is awesome! I'll wait until you rebase it, but want to get it in right away.

Please, though, make sure to separate out the unit tests from any other updates to the code. Put them in separate pull requests.

fpagliughi commented:
The one thing I disagree with, however, is the addition and use of the various `get_c_struct()` methods. Adding them might have two detrimental effects:
(1) Make us (the library writers) lazy, and not provide a full API, if we know it can easily be bypassed.
(2) Allow the users to code to an API that could change out from under us. It would be better if we mitigate any changes to the underlying C library.

Normally I would insist that a library API is complete enough that unit tests can do full testing using only the public API exposed by the library. But in this particular case of a wrapper lib that is closely tied to its underlying C library, perhaps it's a good idea that the unit tests will peek into the C structures and validate them. 

But a better way might be for each class in the library assume that there will be a unit test class that can peek into it. If, by convention, we name the unit test class <class>_test, then we can make it a friend, like:

    class connect_options_test;

    class connect_options
    {
        friend connect_options_test;
        // ...
    };

Would that work?

guilhermeferreira commented:
Branch updated! 

Changes:
- Rebased onto the most recent `develop` branch.
- Isolated the unit tests from other updates to the code. Moved other code changes to individual pull requests.
- Replaced the `get_c_struct()` by a `friend`ship between the unit test class and the class under test (e.g. `connect_options_test` and `connect_options`).
- Placed the unit tests into the `mqtt` namespace.

Open pull requests required to COMPILE/BUILD the tests:
- #45. Prepares the `topic` class to be tested. It allows the use of the `dummy_async_client`.
- #58. Prepares the `ssl_options` class to be tested. It changes the move constructor/assignment semantics and adds an argument constructor.

Open pull requests required to PASS all the tests:
- #41. Fix an invalid memory reference that crashes the tests.
- #49. Evaluating a better approach to fixing the issue.
- #50. Evaluating a better approach to fixing the issue.

Pull requests summary:
- #20 - Required to pass - Merged.
- #24 - Not required - Partly merged.
- #41 - Required to pass - Not merged.
- #42 - Not required - Merged.
- #43 - Required to pass - Merged.
- #44 - Not required - Not merged.
- #45 - Required to compile - Not merged.
- #46 - Required to compile - Merged.
- #47 - Not required - Merged.
- #48 - Not required - Not merged.
- #49 - Required to pass - Not merged.
- #50 - Required to pass - Not merged.
- #58 - Required to compile - Not merged.
